app/models/gem_download.rb

```diff
@@ -0,0 +1,69 @@
+class GemDownload < ActiveRecord::Base
+  belongs_to :rubygem
+  belongs_to :version
+
+  def self.count_for_version(id)
+    v = Version.find(id)
+    return 0 unless v
+    download = GemDownload.find_by(rubygem_id: v.rubygem_id, version_id: v.id)
+    if download
+      download.count
+    else
+      0
+    end
+  end
+
+  def self.count_for_rubygem(id)
+    download = GemDownload.find_by(rubygem_id: id, version_id: 0)
+    if download
+      download.count
+    else
+      0
+    end
+  end
+
+  def self.total_count
+    download = GemDownload.find_by(rubygem_id: 0, version_id: 0)
+    if download
+      download.count
+    else
+      0
+    end
+  end
+
+  def self.update_count_by(count, rubygem_id:, version_id: 0)
+    scope = GemDownload.where(rubygem_id: rubygem_id).select("id")
+    scope = scope.where(version_id: version_id)
+    sql = scope.to_sql
+
+    update = "UPDATE #{quoted_table_name} SET count = count + ? WHERE id = (#{sql}) RETURNING *"
+
+    # TODO: Remove this comments, once we move to GemDownload only.
+    # insert = "INSERT INTO #{quoted_table_name} (rubygem_id, version_id, count) SELECT ?, ?, ?"
+    # find_by_sql(["WITH upsert AS (#{update}) #{insert} WHERE NOT EXISTS (SELECT * FROM upsert)", count, rubygem_id, version_id, count]).first
+    find_by_sql([update, count]).first
```

Comment:
For now, we will only update if the record exists. so we can run redis counts in parallel. 
However, after we change to this, we can insert a GemDownload record, if the record is not there, then we would not need a after_create on rubygem and version.